lib/rust/performance-logger/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+    pub fn release(mut self) {
+        self.released = true;
+        drop(self)
+    }
+}
+
+impl Drop for IntervalHandle {
+    fn drop(&mut self) {
+        if !self.released {
+            WARNING!(format!("{} was dropped without a call to `measure`.", self.interval_name));
+        }
+    }
+}
+
+/// Result of profiling a closure via `measure_interval`. Contains the measurement result and the
+/// closure return value.
+#[derive(Clone, Debug)]
+pub struct IntervalMeasurementResult<T> {
+    /// Return value of the measured closure.
+    pub value:       T,
+    /// Measurement result.
+    pub measurement: Result<Measurement, MeasurementError>,
+}
+
+/// Measure the execution time of the given interval. The interval is executed and the return value
+/// and the measurement result are returned in the `IntervalMeasurementResult`.
+pub fn measure_interval<T, F: FnMut() -> T>(
+    log_level: LogLevel,
+    interval_name: &str,
+    mut closure: F,
+) -> IntervalMeasurementResult<T> {
+    start_interval(log_level, interval_name);
+    let value = closure();
+    let measurement = end_interval(log_level, interval_name);
+
+    IntervalMeasurementResult { value, measurement }
+}
+
+
+
+// ===============
+// === Reports ===
+// ===============
+
+/// Report of all measurements taken during the current session.
+#[derive(Clone, Debug)]
+pub struct Report {
+    measures: Vec<Measurement>,
+}
+
+impl Report {
+    /// Create a new report from the measurements registered in the Performance API.
+    pub fn generate() -> Self {
+        let js_measures = performance().get_entries_by_type("measure");
+
+        let mut measures = Vec::default();
+
+        js_measures.for_each(&mut |entry, _, _| {
+            let entry: PerformanceEntry = entry.into();
+            let measurement: Result<Measurement, _> = entry.try_into();
+            if let Ok(measurement) = measurement {
+                measures.push(measurement)
+            }
+        });
+        Report { measures }
+    }
+}
+
+impl Display for Report {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let sorted_measurements =
+            self.measures.iter().sorted_by_key(|measure| OrderedFloat(measure.start_time));
+        for to_print in sorted_measurements {
+            f.write_str(&format!("{}\n", &to_print))?;
+        }
+        Ok(())
+    }
+}
+
+
+
+// ===============================
+// === Convenience Definitions ===
+// ===============================
+
+macro_rules! define_start_interval_helpers {
+    ($log_level:expr, $log_level_name:ident) => {
+        paste::paste! {
+            #[doc = "Start measuring an interval with log level `" $log_level_name "`. See `start_interval` for more information."]
+            pub fn  [< start_interval_ $log_level_name >](interval_name: &str) -> IntervalHandle {
+                start_interval($log_level, interval_name)
+            }
```

Comment:
Please do not use `paste` macros. They break all IntelliJ hints.

Reply:
Unpasted.